src/python/WMCore/WMSpec/StdSpecs/StdBase.py

```diff
@@ -78,43 +78,60 @@ def __call__(self, workloadName, arguments):
 
         return
 
+    # static copy of the skim mapping
+    skimMap = {}
+
     @staticmethod
-    def skimToDataTier():
+    def skimToDataTier(cmsswVersion, skim):
         """
-        Map physics skim to a data tier
+        Start subprocess and call CMSSW python code to retrieve data tier for given skim
+
+        Detects a usable scram arch for this CMSSW release on this machine
+
+        Cache the results and use cache for lookup if possible
+
         """
-        skimMap = {'LogError': 'RAW-RECO',
-                   'LogErrorMonitor': 'USER',
-                   'ZElectron': 'RAW-RECO',
-                   'ZMu': 'RAW-RECO',
-                   'MuTau': 'RAW-RECO',
-                   'TopMuEG': 'RAW-RECO',
-                   'EcalActivity': 'RAW-RECO',
-                   'CosmicSP': 'RAW-RECO',
-                   'CosmicTP': 'RAW-RECO',
-                   'ZMM': 'RAW-RECO',
-                   'Onia': 'RECO',
-                   'HighPtJet': 'RAW-RECO',
-                   'D0Meson': 'RECO',
-                   'Photon': 'AOD',
-                   'ZEE': 'AOD',
-                   'BJet': 'AOD',
-                   'OniaCentral': 'RECO',
-                   'OniaPeripheral': 'RECO',
-                   'SingleTrack': 'AOD',
-                   'MinBias': 'AOD',
-                   'OniaUPC': 'RAW-RECO',
-                   'HighMET': 'RECO',
-                   'BPHSkim': 'USER',
-                   'PAMinBias': 'RAW-RECO',
-                   'PAZEE': 'RAW-RECO',
-                   'PAZMM': 'RAW-RECO',
-                   'EXONoBPTXSkim': 'USER'
-                   }
-        return skimMap
+        if not cmsswVersion:
+            return None
+
+        if cmsswVersion in StdBase.skimMap:
+            if skim in StdBase.skimMap[cmsswVersion]:
+                return StdBase.skimMap[cmsswVersion][skim]
+        else:
+            StdBase.skimMap[cmsswVersion] = {}
+
+        import glob
+        import subprocess
+
+        p = subprocess.Popen("/cvmfs/cms.cern.ch/common/cmsos", stdout=subprocess.PIPE, shell=True)
+        cmsos = p.communicate()[0].strip()
+
+        scramBaseDirs = glob.glob("/cvmfs/cms.cern.ch/%s*/cms/cmssw/%s" % (cmsos, cmsswVersion))
+        if not scramBaseDirs:
+            scramBaseDirs = glob.glob("/cvmfs/cms.cern.ch/%s*/cms-patch/cmssw/%s" % (cmsos, cmsswVersion))
+            if not scramBaseDirs:
+                return None
+
+        command = "source /cvmfs/cms.cern.ch/cmsset_default.sh\n"
+        command += "cd %s\n" % scramBaseDirs[0]
+        command += "eval `scramv1 runtime -sh`\n"
+
+        command += """python -c 'from Configuration.StandardSequences.Skims_cff import getSkimDataTier\n"""
+        command += """dataTier = getSkimDataTier("%s")\n""" % skim
+        command += """if dataTier:\n\tprint dataTier.value()'"""
+
+        p = subprocess.Popen(command, stdout=subprocess.PIPE, shell=True)
+        dataTier = p.communicate()[0].strip()
```

Comment:
No, if the skim exists in the cmssw version, it should return a valid string. Only if the skim doesn't exist you get "None" (and you get that because at some point upstream it casts the response to a string, which is ugly, but nothing I can do about it at this point).

Hm, about the caching, you are correct. In principle I could make it so the response is cached before the "None" check. It's a bit inconsistent as it is since a real None response would be cached.